File: Python/PROGRAMMERS/STACK QUEUE/다리를지나는트럭.py
def solution(bridge_length, weight, truck_weights):
    Q = [0] * bridge_length
    cur_truck = 0
    time = 0
    while cur_truck != len(truck_weights):
        time+=1
        Q.pop(0)
        if sum(Q) + truck_weights[cur_truck] <= weight:
            Q.append(truck_weights[cur_truck])
            cur_truck+=1
        else:
            Q.append(0)
    return time+bridge_length


bridge_length = 2
weight = 10
truck_weights = [7,4,5,6]
# truck_weights = [10,10,10,10,10,10,10,10,10,10]
solution(bridge_length, weight, truck_weights)

File: JavaScript/PROGRAMMERS/다리를지나는트럭.js
function solution(bridge_length, weight, truck_weights) {
    
    function setArray(arrayLength) {
        const Q = []
        for(let i=0;i<arrayLength;i++) {
            Q.push(0)
        }
        return Q
    }
    
    let idx = 0;
    const Q = setArray(bridge_length);
    let sumOfQ = 0;
    let time = 0;
    
    while (idx<truck_weights.length) {
        const firstElement = Q.shift()
        sumOfQ -= firstElement
        const cur = truck_weights[idx]
        if (cur + sumOfQ <weight) {
            Q.push(cur)
            sumOfQ += cur
            idx += 1
        }
        else {
            Q.push(0)
        }
        time += 1
    }
    return time+bridge_length
}

console.log(solution(2,10,	[7,4,5,6]))